File: solutions/2026-05-10/python/2026-05-10-maximum-number-of-jumps-to-reach-the-last-index.py
```python
# Problem: Maximum Number of Jumps to Reach the Last Index
# Link: https://leetcode.com/problems/maximum-number-of-jumps-to-reach-the-last-index/
#
# Approach:
# This problem can be solved using dynamic programming. We want to find the maximum
# number of jumps to reach each index. Let dp[i] be the maximum number of jumps
# to reach index i.
#
# Base case: dp[0] = 0 (we start at index 0 with 0 jumps).
#
# Recurrence relation:
# To calculate dp[j], we consider all previous indices i (0 <= i < j).
# If we can jump from index i to index j (i.e., -target <= nums[j] - nums[i] <= target),
# and if index i is reachable (dp[i] != -1), then we can potentially reach index j
# from index i. The number of jumps to reach j through this path would be dp[i] + 1.
# We want to maximize dp[j], so we take the maximum of all such possible jumps:
# dp[j] = max(dp[j], dp[i] + 1) for all valid i.
#
# Initialization: Initialize all dp values to -1 (representing unreachable) except for dp[0].
#
# The final answer will be dp[n-1].
#
# Time Complexity Analysis:
# We have a nested loop. The outer loop iterates from j = 1 to n-1 (n-1 iterations).
# The inner loop iterates from i = 0 to j-1 (up to n-1 iterations).
# Inside the inner loop, operations are O(1).
# Therefore, the overall time complexity is O(n^2), where n is the length of the nums array.
#
# Space Complexity Analysis:
# We use a DP array of size n to store the maximum jumps to reach each index.
# Therefore, the space complexity is O(n).
#

class Solution:
    def maximumJumps(self, nums: list[int], target: int) -> int:
        n = len(nums)
        # Initialize dp array. dp[i] will store the maximum number of jumps to reach index i.
        # Initialize with -1, indicating unreachable, except for the starting index.
        dp = [-1] * n
        dp[0] = 0  # We start at index 0 with 0 jumps.

        # Iterate through each index j from 1 to n-1.
        for j in range(1, n):
            # For each index j, iterate through all previous indices i.
            for i in range(j):
                # Check if we can jump from index i to index j.
                # The condition is -target <= nums[j] - nums[i] <= target.
                if -target <= nums[j] - nums[i] <= target:
                    # If index i is reachable (dp[i] != -1), then we can potentially reach index j.
                    if dp[i] != -1:
                        # Update dp[j] with the maximum number of jumps.
                        # If dp[j] was -1, this is the first way to reach j.
                        # If dp[j] was not -1, we are trying to find a path with more jumps.
                        dp[j] = max(dp[j], dp[i] + 1)

        # The result is the maximum number of jumps to reach the last index (n-1).
        # If dp[n-1] is still -1, it means the last index is unreachable.
        return dp[n - 1]

```

File: solutions/2026-05-10/javascript/2026-05-10-maximum-number-of-jumps-to-reach-the-last-index.js
/**
 * @summary Calculates the maximum number of jumps to reach the last index in an array.
 * Jumps are restricted by the difference between array elements and a given target value.
 * @link https://leetcode.com/problems/maximum-number-of-jumps-to-reach-the-last-index/
 * @approach This problem can be solved using dynamic programming. We define dp[i] as the maximum number of jumps to reach index i.
 * The base case is dp[0] = 0, as we start at index 0 with 0 jumps.
 * For each index i from 1 to n-1, we iterate through all previous indices j (0 <= j < i).
 * If a jump from index j to index i is valid (i.e., -target <= nums[i] - nums[j] <= target),
 * and if index j is reachable (dp[j] != -1), we can potentially reach index i from index j.
 * We update dp[i] with the maximum of its current value and dp[j] + 1.
 * If after checking all possible previous indices, dp[i] remains -1, it means index i is unreachable.
 * Finally, dp[n-1] will contain the maximum number of jumps to reach the last index, or -1 if unreachable.
 *
 * Time Complexity: O(n^2) where n is the length of the nums array.
 * The outer loop iterates from i = 1 to n-1, and the inner loop iterates from j = 0 to i-1.
 *
 * Space Complexity: O(n) where n is the length of the nums array.
 * We use a DP array of size n to store the maximum jumps to reach each index.
 */
const maximumJumps = (nums, target) => {
    const n = nums.length;
    // dp[i] will store the maximum number of jumps to reach index i.
    // Initialize all dp values to -1, indicating that indices are unreachable by default.
    const dp = new Array(n).fill(-1);
    // Base case: We start at index 0 with 0 jumps.
    dp[0] = 0;

    // Iterate through each index from 1 to n-1 to calculate the maximum jumps to reach it.
    for (let i = 1; i < n; i++) {
        // For each index i, iterate through all previous indices j.
        for (let j = 0; j < i; j++) {
            // Check if index j is reachable (dp[j] is not -1).
            // Also, check if a jump from index j to index i is valid according to the target condition.
            if (dp[j] !== -1 && Math.abs(nums[i] - nums[j]) <= target) {
                // If a jump is valid and index j is reachable, we can potentially reach index i from j.
                // Update dp[i] with the maximum number of jumps.
                // This is either the current dp[i] value or the jumps to reach j plus one more jump.
                dp[i] = Math.max(dp[i], dp[j] + 1);
            }
        }
    }

    // The result is the maximum number of jumps to reach the last index (n-1).
    // If dp[n-1] is still -1, it means the last index is unreachable.
    return dp[n - 1];
};
```